### ISSUE TITLE
Detailed differences between "reported" and "this code"

### ISSUE THREAD

**Issue by morningmoni** [shmsw25/AmbigQA#5]
Hi Sewon,

I found in the table https://github.com/shmsw25/AmbigQA/tree/master/codes#aggregated-results the performance appears to improve significantly compared to the reported results (e.g., 42.2 -> 45 on NQ-open, 30.8 -> 35.8 on AmbigQA zero-shot). I wonder besides "re-implemented the models with Huggingface Transformers", what other changes were made to achieve such performance boosts (e.g., did you change parameters like train_batch_size, train_M)? Also, if I understand correctly, SpanSeqGen (this code) uses BART-base, which outperforms DPR BERT-large despite fewer parameters, right?

Thank you!

**Comment by shmsw25** [comment]
Thanks for your interest!

**Reimplementation of SpanSeqGen:** The model architecture is identical. The previous version was implemented in fairseq and this version is in HF Transformers, which could give a lot of differences in details. To be honest, I wasn't familiar with fairseq at that time and only changed input-output without modifying the internal codes. That could be reason that the number wasn't so good then. Some more explicit differences I made for this codebase is:
- When there are multiple accepted answer sequences, this codebase samples one answer sequence at each iteration, whereas previous one considered each answer as separate instance.
- Previous one evaluated on the dev & saved checkpoint every one epoch, while this one does it every 500 steps.


**RE BERT vs. BART-large:** In fact, SpanSeqGen uses BART-large. Yes SpanSeqGen slightly outperforms DPR BERT-large (although the difference is so marginal so I would say they are almost the same). I think BART-large has more paramters than BERT-large. (400M vs. 340M).

**Comment by morningmoni** [comment]
Thank you~

**Comment by morningmoni** [comment]
Just to clarify.. in the paper it says "sequentially generates distinct answers token-by-token, separated by [SEP], in the order in which they appear in the input passages." If there are two answers like ['A B', 'C'], it sounds to me the target sequence would be 'A B [SEP] C'? Why do you need to "sample one answer sequence at each iteration" then?

**Comment by shmsw25** [comment]
You are right. For example, if two answers are ["United States" and "2021"], the generated sentence is something like "United States [SEP] 2021". The reason I has "sample one answer sequence at each iteration" is because there are multiple acceptable answers.
For instance, if you look at AmbigQA annotation, each answer is annotated as a list, e.g., [["United States", "US"], ["2021"]]. This means that there are two distinct answers and you would want to generate either of "United States [SEP] 2021" or "US [SEP] 2021". But you would not want to generate "United States [SEP] US [SEP] 2021" because "United States" and "US" refer to one semantically unique answer.
Therefore, in this case, our codebase samples one output sequence at each iteration, either of "United States [SEP] 2021" or "US [SEP] 2021", and uses it to train the model.

Does it make sense to you? Let me know if you want me to clarify anything.

**Comment by morningmoni** [comment]
Yea, that makes sense! For NQ-open, I guess you don't have to sample since the answers is not a nested list (no synonyms)? How do you use the generated answer sequence for evaluation (EM) if it contains **multiple** answers and you want to match **one** of the gold reference answers?

**Comment by shmsw25** [comment]
For NQ-open: it actually also contains multiple acceptable answers. For example, each annotated answer is a list (e.g. ["United States", "US"]), meaning that there is a single answer which can be expressed as "United States" or "US". Thus, we did sampling of an output sequence for NQ-open as well.

Great question about evaluation. If the prediction matches any of gold reference answers, we consider it as correct. (e.g. `em = np.max([exact_match_score(prediction, groundtruth) for groundtruth in groundtruths])`

**Comment by morningmoni** [comment]
Ahh.. let me rephrase. Say if the generated answer is "United States [SEP] Canada", how would you use it as the "prediction" in the code above? 
Or even if the generated answer is "United States [SEP] US", you'll still need to split it (to  ["United States", "US"]?) and select one of them as the prediction to match one answer, right?

**Comment by shmsw25** [comment]
Oh, I got your question. For the exact description, I would refer to the [paper](https://arxiv.org/pdf/2004.10645.pdf) (Sec 3.2) and the [evaluation script](https://github.com/shmsw25/AmbigQA/blob/master/codes/ambigqa_evaluate_script.py#L231), but let me summarize here.

It is fine to predict any of accepted answers when they refer to a single semantically meaningful answer. If you predict a subset of semantically different answers, you get partial credits based on F1 score. For instance, in the above example where the groundtruth is [["United States", "US"], ["2021"]],

Prediction: `United States [SEP] 2021` : Full score (F1 1.0)
Prediction: `US [SEP] 2021` : Full score (F1 1.0)
Prediction: `United States` : F1 0.67 (Prec = 1, Recall = 0.5)
Prediction: `United States [SEP] US` : F1 0.5 (Prec = 0.5, Recall = 0.5; here precision is 0.5 because you've already got 'United States', so 'US' is considered as extra, incorrect prediction. So you shouldn't have predicted US if you've already predicted United States, which is possible only when you know they refer to the same answer.)

**Comment by morningmoni** [comment]
Thanks for the detailed examples! I agree that F1 makes more sense if the question is ambiguous and contains multiple semantically different answers. However, what I was wondering is the EM evaluation on NQ-open (EM=45 in the table, not on AmbigQA with additional annotations). For NQ-open, if "groundtruths" is a list of answers like ["United States", "US"], and the "predictions" is a list of answers ["United States", "Canada"] instead of just one answer as in DPR, what is its EM score? What I don't understand is which answer to take to measure EM: if you take "United States" then it's exactly matched and if you take "Canada" then it's not.

**Comment by shmsw25** [comment]
Got it, so for NQ-open evaluation, we did not use AmbigQA annotations at all, and only used the original annotations of NQ-open. This is because we wanted our numbers to be comparable to previous works.
These NQ-open annotations can be found [here](https://github.com/efficientqa/nq-open). You can see that answers are annotated as lists, which is consistent to my description earlier:

> For NQ-open: it actually also contains multiple acceptable answers. For example, each annotated answer is a list (e.g. ["United States", "US"]), meaning that there is a single answer which can be expressed as "United States" or "US". Thus, we did sampling of an output sequence for NQ-open as well.

**Comment by morningmoni** [comment]
Got it, so it's the same as in [Fusion-in-Decoder](https://arxiv.org/pdf/2007.01282.pdf) then. Thanks!

**Comment by shmsw25** [comment]
Exactly, it should be the same as Fusion-in-Decoder, or any other models that use sequence-to-sequence model as its reader. Thanks for asking :)
